doc/tools/docgram/orderedGrammar

```diff
@@ -792,17 +792,17 @@ reference: [
 | string OPT [ "%" scope_key ]
 ]
 
-argument_spec: [
-| OPT "!" name OPT ( "%" scope_key )
-]
-
 arg_specs: [
 | argument_spec
+| "(" LIST1 argument_spec ")" LIST0 ( "%" scope )
+| "[" LIST1 argument_spec "]" LIST0 ( "%" scope )
+| "{" LIST1 argument_spec "}" LIST0 ( "%" scope )
```

Comment:
Please keep these lines after `"&"` as they were, also use that order in the description of the Arguments command.  Let's not change it without a reason.

Reply:
Well I explained my rationale above: https://github.com/coq/coq/pull/16472#discussion_r969319803 you can disagree with it, we can discuss it but ignoring it feels a bit rude. As I said, there is the following list just below the grammar of the `Arguments` command:

> * Making arguments implicit. Afterward, implicit arguments
>     must be omitted in any expression that applies :token:`reference`.
> * Declaring that some arguments of a given function should
>     be interpreted in a given scope.
> * Affecting when the :tacn:`simpl` and :tacn:`cbn` tactics unfold the function.
>     See :ref:`Args_effect_on_unfolding`.
> * Providing bidirectionality hints.  See :ref:`bidirectionality_hints`.

The first item correspond to all the `@name`, `(...)`, `[...]`and `{...}`, the second one to the `%scope` things, the third one to `/` and `!` while the last one corresponds to `&`. Hence the order I'd like to follow (but maybe we could do better).

Reply:
I think you initially missed a few of my earlier comments.  Should I call that rude?  I think not.  It's very easy to miss details, plus it's so easy to misinterpret electronic communications.  It's better just to raise a point again.  Your comments here have more detail that you didn't give earlier.

The following changes are inconsistent with what we've done elsewhere in the manual:
- Text describing grammar items should appear in exactly the order they appear in the grammar so it's easy to find.  So the text for ``` :n:`@name {? % @scope }` ``` is in the wrong place.  The grammar is sorted by dependencies, so `argument_spec` should be after `arg_specs`.
- The entry for `!` is in the wrong place compared to the grammar
- It's not essential that the bullet items you quote are in exactly the order matching the grammar.  The correspondence between the bullet items and the syntax descriptions is neither explicit in the text nor obvious (at least not to an expert).

The text doesn't say that `@name`, `(...)` is about implicit parameters at all.  How does it work?

Also I'd note that the description of the command is pretty weak; it doesn't tell you how to use the command.  And it is a bit of a hodge podge.  (That's not a request to improve it in this PR!)

We can consider whether to move `/` and `&` as you have.

Reply:
Thanks for the explanation. So indeed you are right that I should revert to the original order (except for `argument_spec` that should be after `arg_specs` in the grammar). What about the following:

1. reverting everything to original order
1. except for `argument_spec` that should come just after `arg_specs`
1. put the two first items of the bullet list last, to better match the order of the grammar (although this order doesn't have to match, I still feel it easier to read when it is more consistent)
2. this two first bullet should link to the parts of the manual about implicit arguments and scopes respectively

Regarding the entry for `!` I understand its position according to the grammar but being semantically much closer from `/`, it feels a bit strange when reading it. Shouldn't that warrant an exception to the rule? (`/` and `!` are syntactically in different places in the grammar because the first must appear only one while the last can be attached to any argument but they are both about the very same thing in practice) 

> The text doesn't say that @name, (...) is about implicit parameters at all. How does it work?

Indeed, this is only specified in the ``:n:`name`` entry 

> The construct :n:`@name {? % @scope }` declares :n:`@name` as non-implicit if `clear implicits` is spec\
ified or at least one other name is declared implicit in the same list of :n:`@name`\s.

Maybe this sentence should be copied or moved to the ``:n:`( ... )`` entry?

> Also I'd note that the description of the command is pretty weak; it doesn't tell you how to use the command. And it is a bit of a hodge podge. (That's not a request to improve it in this PR!)

There are examples in the subsections below, but maybe they need improvement (in a further PR indeed)

> We can consider whether to move / and & as you have.

Well, now that I understand the rules, I feel the original order is just as good.

Reply:
I agree on 1 and 2 above.

>> We can consider whether to move / and & as you have.

> Well, now that I understand the rules, I feel the original order is just as good.

"moving" meaning changing the order both in the grammar and the text.  That would make them adjacent to `!`, which seems in line with your thinking.

> 3 put the two first items of the bullet list last

I don't see the benefit.  The bullet list is meant to be in descending order of importance/frequency of use.  IINM the order already follows that principle.

> 4 this two first bullet should link to the parts of the manual about implicit arguments and scopes respectively

In the first, maybe change "implicit arguments" to ```:term:`implicit arguments` ```.  (But note the implicit arguments chapter is not easy to read, either.  It needs an initial section introducing the main concepts with several short examples.  I think examples are usually easier for readers to grasp and the text is more for elaboration.  Hence the popularity of cut and paste.)

In the second, create a new glossary term by using ```:gdef:`scope` ``` at the target point of the link and then ```:term:`scope` ``` to refer to it.

> Maybe this sentence should be copied or moved to the ``:n:`( ... )`` entry?

Eh.  The sentence you refer to is hardly a model of clarity or completeness.  There should be a paragraph or two here with short examples explaining the "implicit" concept separate from the text describing the syntax .  A better, more succinct sentence could be repeated in multiple entries; in a reference manual, you want to minimize the amount of text that has to be read to understand something.  OTOH, we don't want to needlessly repeat somewhat complex points.  I would leave it alone for now.

Reply:
> I agree on 1 and 2 above.

Done
I also took the opportunity to do a squash and rebase to clean the log.
 
> > > We can consider whether to move / and & as you have.
> 
> > Well, now that I understand the rules, I feel the original order is just as good.
> 
> "moving" meaning changing the order both in the grammar and the text. That would make them adjacent to `!`, which seems in line with your thinking.

Right, but then the `name` entry gets separated from the `( ... )`, `[ ... ]` and `{ ... }` ones, so that may not be really better. Let's stick with  the original order.
 
> > 3 put the two first items of the bullet list last
> 
> I don't see the benefit. The bullet list is meant to be in descending order of importance/frequency of use. IINM the order already follows that principle.

Ok, you're right.

> > 4 this two first bullet should link to the parts of the manual about implicit arguments and scopes respectively
> 
> In the first, maybe change "implicit arguments" to `` :term:`implicit arguments` ``.

Done (changing plural to singular as it's the way the term was defined)

> (But note the implicit arguments chapter is not easy to read, either. It needs an initial section introducing the main concepts with several short examples. I think examples are usually easier for readers to grasp and the text is more for elaboration. Hence the popularity of cut and paste.)

Ouch, indeed that section would need some major work to make it more accessible. It directly dives into technicality of how automatic inference of implicit arguments work without really explaining what implicit arguments are in the first place.
 
> In the second, you could create a new glossary term by using `` :gdef:`scope` `` at the target point of the link and then `` :term:`scope` `` to refer to it.

Thanks, done
 
> > Maybe this sentence should be copied or moved to the `` :n:`( ... ) `` entry?
> 
> Eh. The sentence you refer to is hardly a model of clarity or completeness. There should be a paragraph or two here with short examples explaining the "implicit" concept separate from the text describing the syntax . A better, more succinct sentence could be repeated in multiple entries; in a reference manual, you want to minimize the amount of text that has to be read to understand something. OTOH, we don't want to needlessly repeat somewhat complex points. I would leave it alone for now.

Ok, let's defer that to a further PR (the one on implicit arguments might be more urgent anyway).

Reply:
> > In the first, maybe change "implicit arguments" to `` :term:`implicit arguments` ``.

> Done (changing plural to singular as it's the way the term was defined)

```:term:`implicit arguments <implicit argument>` ``` will show the plural with the correct link

Reply:
Indeed, thanks